lib/engine-object-file/src/artifact.rs

```diff
@@ -0,0 +1,647 @@
+        - Features
+        - ModuleInfo
+        - MemoryIndex -> MemoryStyle
+        - TableIndex -> TableStyle
+        - LocalFunctionIndex -> FunctionBodyPtr // finished functions
+        - FunctionIndex -> FunctionBodyPtr // finished dynamic function trampolines
+        - SignatureIndex -> VMSharedSignatureindextureIndex // signatures
+         */
+
+        let serialized_data = bincode::serialize(&metadata).map_err(to_compile_error)?;
+        let mut metadata_binary = vec![0; 10];
+        let mut writable = &mut metadata_binary[..];
+        leb128::write::unsigned(&mut writable, serialized_data.len() as u64)
+            .expect("Should write number");
+        metadata_binary.extend(serialized_data);
+        let metadata_length = metadata_binary.len();
+
+        let maybe_obj_bytes = compiler.experimental_object_file_compile_module(
+            &target,
+            &metadata.compile_info,
+            module_translation.as_ref().unwrap(),
+            &function_body_inputs,
+            &metadata,
+            &metadata_binary,
+        );
+
+        let obj_bytes = if let Some(obj_bytes) = maybe_obj_bytes {
+            obj_bytes?
+        } else {
+            let compilation = compiler.compile_module(
+                &target,
+                &metadata.compile_info,
+                module_translation.as_ref().unwrap(),
+                function_body_inputs,
+            )?;
+            let mut obj = get_object_for_target(&target_triple).map_err(to_compile_error)?;
+            emit_data(&mut obj, WASMER_METADATA_SYMBOL, &metadata_binary)
+                .map_err(to_compile_error)?;
+            emit_compilation(&mut obj, compilation, &metadata, &target_triple)
+                .map_err(to_compile_error)?;
+            obj.write().map_err(to_compile_error)?
+        };
+
+        //let host_target = Triple::host();
+        //let is_cross_compiling = target_triple != &host_target;
+
+        Self::from_parts_crosscompiled(&mut *engine_inner, metadata, obj_bytes, metadata_length)
+    }
+
+    /// Generate the header file that goes with the generated object file.
+    pub fn generate_header_file(&self) -> String {
+        use crate::header::*;
+        let mut c_statements = vec![];
+        c_statements.push(CStatement::Declaration {
+            name: "module_bytes_len".to_string(),
+            array: false,
+            is_extern: false,
+            is_const: true,
+            ctype: CType::U32,
+            definition: Some(Box::new(CStatement::LiteralConstant {
+                value: format!("{}", self.metadata_length),
+            })),
+        });
+        c_statements.push(CStatement::Declaration {
+            name: "WASMER_METADATA".to_string(),
+            array: true,
+            is_extern: true,
+            is_const: true,
+            ctype: CType::U8,
+            definition: None,
+        });
+        for (function_local_index, _function_len) in self.metadata.function_body_lengths.iter() {
```

Comment:
Walking through body lengths to use the function local index only seems a bit of misuse.

Reply:
I agree, I copied this code from native engine though haha